### PR TITLE
Add macros %{ocaml_flags:...} and %{link_flags:...}

### PR DESCRIPTION
**Pull request by nojb**
This PR adds two macros that help writing explicit rules in some cases where the extra flexibility is needed. The first one,

- `%{ocaml_flags:mode}` gives the ocaml flags as computed by `dune` in the current directory.

- `%{link_flags:mode:libs}` produces the list of arguments needed to link an executable with `libs` (comma separated list of lib names).

(`mode` is either `byte` or `native`)

As mentioned, these are not for the casual user, but it gives an escape hatch when needed.

In my use case, I need to create `cmxs` and `exe`'s sharing a large number of modules (plugins) in a way that doesn't lend itself to a stratification by libraries and where copying modules around is too inefficient and just not convenient. With these variables I can define a "big" library that includes all plugins in order to set up the compilation of individual modules and can define explicit rules for the linking commands producing the individual `cmxs`'s and `exe`'s in a way "compatible" with `dune`.

It is not beautiful, but it is a fairly specialized need and I don't see an easy way to express this in `dune`'s current model.

Opinions?

### CODE REVIEW

**Comment by rgrinberg** [comment]
We are planning to add per module/file flags to dune. How would these macros in conjunction with such features?

**Comment by nojb** [comment]
These macros produce the flags in the *current* directory; I think it is a notion that is completely independent of any notion of per-module/per-file flag which happen inside libraries, executables, etc.

**Comment by nojb** [comment]
Good point, will amend.

**Comment by nojb** [comment]
I switched to using simple variables `%{ocamlopt_flags}` and `%{ocamlc_flags}` instead of a macro.

File: src/super_context.ml [hunk 1]
```diff
@@ -217,13 +217,13 @@ module Pkg_version = struct
     Build.vpath spec
 end
 
-let partial_expand sctx ~dep_kind ~targets_written_by_user ~map_exe
+let partial_expand sctx ~dir ~dep_kind ~targets_written_by_user ~map_exe
```
Comment (rgrinberg) [inline src/super_context.ml]: You should be able to use the `dir` from the expander here.

File: src/expander.ml [hunk 2]
```diff
@@ -260,6 +263,40 @@ let expand_and_record acc ~map_exe ~dep_kind ~scope
...
+    in
+    begin match libs with
+    | Ok libs ->
+      List.iter ~f:(fun lib -> Resolved_forms.add_lib_dep acc (Lib.name lib) dep_kind) libs;
+      begin match Lib.closure libs ~linking:true with
+      | Ok libs ->
+        let c_include_flags =
+          Lib.L.c_include_paths libs ~stdlib_dir:t.stdlib_dir |>
+          Path.Set.to_list |>
+          List.map ~f:(fun path -> [Value.String "-I"; Value.Dir path]) |> List.flatten
```
Comment (rgrinberg) [inline src/expander.ml]: We have `List.concat_map`.

File: src/expander.ml [hunk 3]
```diff
@@ -260,6 +263,40 @@ let expand_and_record acc ~map_exe ~dep_kind ~scope
...
+        in
+        let archive_files = Value.L.paths (Lib.L.archive_files libs ~mode) in
+        Some (c_include_flags @ archive_files)
+      | Error exn ->
+        Resolved_forms.add_fail acc { fail = fun () -> raise exn }
+      end
+    | Error exn ->
```
Comment (rgrinberg) [inline src/expander.ml]: Nit: I find it easier to read when the shorter cases come first.

File: src/expander.ml [hunk 4]
```diff
@@ -260,6 +263,40 @@ let expand_and_record acc ~map_exe ~dep_kind ~scope
...
+        (List.map ~f:(fun lib_name -> Lib_name.of_string_exn ~loc:(Some loc)
+                                        lib_name) lib_names)
+    in
+    begin match libs with
+    | Ok libs ->
+      List.iter ~f:(fun lib -> Resolved_forms.add_lib_dep acc (Lib.name lib) dep_kind) libs;
+      begin match Lib.closure libs ~linking:true with
+      | Ok libs ->
+        let c_include_flags =
+          Lib.L.c_include_paths libs ~stdlib_dir:t.stdlib_dir |>
```
Comment (rgrinberg) [inline src/expander.ml]: Why can't we just use `Lib.L.include_flags`?

**Comment by diml** [comment]
I just remembered that `%{env:X}` is for accessing environment variables :/ So `%{env:flags}` might not work indeed. Maybe `%{dune-env:flags}` or `%{env.flags}`?

**Comment by nojb** [comment]
> I just remembered that %{env:X} is for accessing environment variables :/ So %{env:flags} might not work indeed. Maybe %{dune-env:flags} or %{env.flags}?

Indeed, I find it confusing to use `%{env:...}` for this as well (strictly speaking, the environment variable case requires a `=` symbol, but still). I will give `%{dune-env:...}` a try.

**Comment by nojb** [comment]
This has bitrotted. Closing for now, may re-open in the future as needed.

**Pull request closed without merging**